security/providers/oidc/src/main/java/io/helidon/security/providers/oidc/OidcSupport.java

```diff
@@ -242,6 +268,48 @@ private void processLogout(ServerRequest req, ServerResponse res) {
                 .exceptionallyAccept(t -> sendError(res, t));
     }
 
+    private Single<String> findTenantName(ServerRequest request) {
+        List<String> missingLocations = new LinkedList<>();
+        Optional<String> tenantId = Optional.empty();
+        if (oidcConfig.useParam()) {
+            tenantId = request.queryParams().first(oidcConfig.tenantParamName());
+
+            if (tenantId.isEmpty()) {
+                missingLocations.add("query-param");
+            }
+        }
+        if (oidcConfig.useCookie() && tenantId.isEmpty()) {
+            Optional<Single<String>> cookie = oidcConfig.tenantCookieHandler()
+                    .findCookie(request.headers().toMap());
+
+            if (cookie.isPresent()) {
+                return cookie.get();
+            }
+            missingLocations.add("cookie");
+        }
+        if (tenantId.isPresent()) {
+            return Single.just(tenantId.get());
+        } else {
+            if (LOGGER.isLoggable(Level.FINEST)) {
+                LOGGER.finest("Missing tenant id, could not find in either of: " + missingLocations
+                                      + "Falling back to the default tenant id: " + DEFAULT_TENANT_ID);
+            }
+            return Single.just(DEFAULT_TENANT_ID);
+        }
+    }
+
+    private Tenant obtainCurrentTenant(String tenantName) {
```

Comment:
This method is blocking (`Tenant.create()` may invoke HTTP requests against metadata endpoint). This cannot run in web server request as designed here